### DISCUSSION TITLE
`Filter!   diff.ignoreSubmodules=untracked` on top of magit buffer is confusing

### DISCUSSION THREAD

**Discussion by Atemu** [magit/magit#5121] [Ideas]
What is that message trying to tell me?

From reading https://github.com/magit/magit/issues/4019 I gather that this message is intentional. I guess it's to inform the user that Magit may no function correctly with that setting?

AFAICT, setting it to `untracked` does not affect magit in any way because magit ignores this "modified" status already. I originally set this config flag because I also don't care about untracked files in my submodules when I'm outside of magit; CLI git shows submodules with untracked files as "modified" in `git status`.

1. The message should explain itself a little better
2. There should be an obvious `i-know-what-im-doing` variable which disables the message
3. Should this message even show with `diff.ignoreSubmodules=untracked`?

**Comment by tarsius** [comment]
Yes, there's room for improvements along those lines.

For now you can hide this using

```elisp
(remove-hook 'magit-status-headers-hook
             'magit-insert-diff-filter-header)
```

**Reply by Atemu** [reply]
Does `diff.ignoreSubmodules=untracked` inhibit any of Magit's functionality?

**Reply by tarsius** [reply]
When I took a quick look, I got the impression that it does not.

Dealing with the various things you have brought up, will be a series of (hopefully) small steps, and after each one awaits another. I decided to stop here. I am currently focused on other things and it will be a while until I get around to dealing with such details again.

Also note that since `untracked` is the default (unless I am mistaken), you too can do something to prevent that from being shown; just don't double down on the default! I *believe* that shouldn't make a difference, except with regards to the `Filter!` line.

**Reply by Atemu** [reply]
Sure, thanks a bunch for your work :)

Contrary to what the git-diff man page suggests, `untracked` does not appear to be the default in `git status`:

```
$ git status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
        modified:   test (untracked content)

no changes added to commit (use "git add" and/or "git commit -a")
$ git -c diff.ignoreSubmodules=untracked status
On branch master
nothing to commit, working tree clean
```

**Comment by tarsius** [comment]
I've added help text and the status buffer now shows modules with untracked content by default now. cf21b6163fbb392bc3d708767749b1619e644138~..d7e0ee3a3829e056faf886ba5186545f87206870

**Reply by Atemu** [reply]
Am I reading this correctly that it now overrides the git user setting? I have set that option precisely because I want submodules with untracked files to not show up as dirty/changed in git status.

I'd prefer if magit just behaved the same as vanilla git here.

**Reply by tarsius** [reply]
That's the default now, yes. (Another user asked for that, and I wanted that too.) That default is established by using the diff argument `--ignore-submodules=none`. Such arguments can be changed and the choice then saved. In a status buffer press `D`, change the argument and then save. You might want to do the same in a diff buffer (whose arguments are set separately from those of status buffers). https://docs.magit.vc/transient/Saving-Values.html

**Reply by Atemu** [reply]
Hm, I don't quite follow why magit needs to override this setting at all by default? Your preferences are yours of course (and it's your project obviously) but couldn't/shouldn't that preference be expressed by simply setting `diff.ignoreSubmodules`? I think what I'm asking is: Is there a specific reason why magit must deviate from the git CLI w.r.t. this preference/option?

**Reply by tarsius** [reply]
`diff.ignoreSubmodules` affects diff and status buffers, but it makes sense for this to be customizable per buffer-type. Furthermore the documentation about this variable in git-diff(1) is incorrect, the default it mentions is not actually the default. It also fails to mention that it affects `git status` (where the default is not the same as for `git diff`, if I remember correctly).

By the way, you don't have to set a value for this argument, you can also leave it undefined, in which case the value of the variable will be used.

**Reply by Atemu** [reply]
> it makes sense for this to be customizable per buffer-type

Certainly. And also dynamically at runtime. I do not intend to argue against magit *allowing* the user to deviate from the default; merely that it should not do so by default.

> the documentation about this variable in git-diff(1) is incorrect, the default it mentions is not actually the default.

That is true but that's an upstream git issue, not magit, and the fix would probably be to update the documentation to follow the behaviour, not the other way around.

> By the way, you don't have to set a value for this argument, you can also leave it undefined, in which case the value of the variable will be used.

Sorry, I'm not following; which variable? The git config one?

The diff reads to me as though magit provides a default setting that overrides the "normal" git behaviour (as decided by git config) which implies to me that the git config setting would be ineffective unless I specifically tell magit not to override it (by modifying `magit-diff-default-arguments`). I might be misunderstanding though.

**Reply by tarsius** [reply]
I am going on vacation now. Please try to live with my decision.
